### ISSUE TITLE
Texture issue?

### ISSUE THREAD

**Issue by EpicLPer** [Sage-of-Mirrors/SuperBMD#4]
Heya,

so I've been trying to export models from Super Mario Galaxy with various different programs now. When I try to export some of them the textures however have the wrong color, is that due to the program not doing something right or because Nintendo using shaders to color those objects instead of textures?

Here's how it should look like:
![discord_2018-01-27_04-08-53](https://user-images.githubusercontent.com/8442196/35472945-e01e82c8-0378-11e8-9727-4b332dfa92c0.png)

But here's how SuperBMD and other programs actually export it:
![blender_2018-01-27_15-38-52](https://user-images.githubusercontent.com/8442196/35472946-ea36fc04-0378-11e8-8dc5-ff171fce9448.png)

Is there any way to fix this? Thanks!

**Comment by Sage-of-Mirrors** [comment]
Hey! Can you upload the .bdl file for me so I can take a look at it?

**Comment by EpicLPer** [comment]
Here you go:
[warppod.zip](https://github.com/Sage-of-Mirrors/SuperBMD/files/1670421/warppod.zip)

**Comment by EpicLPer** [comment]
Okay so it does seem to be shaders that do the coloring... There is an option in BDMViewer2 to disable GLSL shaders and when I do that it looks like the exported model (with transparent textures tho)

So unless there's a way to "bake shaders" onto a model or port them somehow over to Unity (which I'd need those to work in then) there's probably no way and it's not a bug :( Sad for me tho...

![bmdview2_2018-01-27_19-50-29](https://user-images.githubusercontent.com/8442196/35475263-8b3bfb8c-039b-11e8-8912-41ddd5db3619.png)

**Comment by LordNed** [comment]
Have you had a look at it in our [J3D Model Viewer?](https://github.com/LordNed/J3D-Model-Viewer)

Nintendo often uses textures to control effects, but they can also set up to 8 color values when they render the object (from the actual CPU code in game) which lets them shift colors over time and do other effects caused by code.

In J3D Model Viewer you can at least override the colors that are being sent to the GPU for the model and place around with their values to see if they change anything on the resulting model. The other thing you can do is that J3D Model Viewer, by default (iirc) dumps the textures and GLSL shader code for the models it can render.

You could take these dumped shaders and re-write them in Unity's CG language (they're pretty similar) which would let you set up a Unity shader that had similar behavior, and you could expose the 8 color values the GameCube can use as color parameters for a Unity shader. You'll need to know Unity's shader stuff, can't say I really recommend porting them unless you really want it to act the same, it'd really be easier to just re-create the effect from scratch to do the same thing.

**Comment by EpicLPer** [comment]
Yeah, as mentioned in the last picture they do it with shaders from what I see. Not sure how easy it'd be to port those over, I'm not a programmer myself let alone when it comes to low-level stuff like shaders :(

I may have to just re-color the things I can see inside Unity as far as possible I guess. Because I don't think there's a way to include an option like "bake colors onto textures" or something?

**Comment by LordNed** [comment]
Yeah I'm certainly not good enough at shaders to figure out a way to bake anything into a resulting texture due to the complexity involved in rendering GC models (ie: they can use a texture map to change the UVs and sample that as another texture map, and then twist and turn it through many stages performing operations where the output of the last stage is the input to the next).
